### PR TITLE
Refactoring `finalize` command

### PR DESCRIPTION
**Pull request by 0xjei**

### CODE REVIEW

File: packages/actions/src/helpers/utils.ts [hunk 1]
```diff
@@ -220,10 +220,17 @@ export const getContributionsValidityForContributor = async (
...
-export const getPublicAttestationPreambleForContributor = (contributorIdentifier: string, ceremonyName: string) =>
-    `Hey, I'm ${contributorIdentifier} and I have contributed to the ${ceremonyName} MPC Phase2 Trusted Setup ceremony.\nThe following are my contribution signatures:`
+export const getPublicAttestationPreambleForContributor = (
+    contributorIdentifier: string,
+    ceremonyName: string,
+    isFinalizing: boolean
+) =>
+    `Hey, I'm ${contributorIdentifier} and I have ${
+        isFinalizing ? "finalized" : "contributed"
+    } to the ${ceremonyName} MPC Phase2 Trusted Setup ceremony.\nThe following are my contribution signatures:`
```
Comment (ctrlc03) [inline packages/actions/src/helpers/utils.ts:233]: move "to" with "contributed"

File: packages/backend/src/functions/circuit.ts [hunk 2]
```diff
@@ -678,3 +681,81 @@ export const refreshParticipantAfterContributionVerification = functionsV1.fires
...
+
+        if (!ceremonyDoc.data() || !circuitDoc.data() || !participantDoc.data() || !contributionDoc.data())
+            logAndThrowError(COMMON_ERRORS.CM_INEXISTENT_DOCUMENT_DATA)
+
+        // Extract data.
+        const { prefix: circuitPrefix } = circuitDoc.data()!
+        const { files } = contributionDoc.data()!
+
+        // Prepare filenames and storage paths.
+        const verificationKeyFilename = `${circuitPrefix}_vkey.json`
```
Comment (ctrlc03) [inline packages/backend/src/functions/circuit.ts:715]: commonTerms?

File: packages/backend/src/functions/circuit.ts [hunk 3]
```diff
@@ -678,3 +681,81 @@ export const refreshParticipantAfterContributionVerification = functionsV1.fires
...
+        const verifierContractFilename = `${circuitPrefix}_verifier.sol`
```
Comment (ctrlc03) [inline packages/backend/src/functions/circuit.ts:716]: as above

File: packages/phase2cli/src/lib/prompts.ts [hunk 4]
```diff
@@ -515,35 +515,42 @@ export const promptZkeyGeneration = async (): Promise<boolean> => {
...
         })
-    }
 
     // Prompt for selection.
     const { ceremony } = await prompts({
         type: "select",
         name: "ceremony",
-        message: theme.text.bold("Which ceremony would you like to contribute to?"),
+        message: theme.text.bold(
+            !isFinalizing ? "Which ceremony would you like to contribute to?" : "What ceremony do you want to finalize?"
```
Comment (ctrlc03) [inline packages/phase2cli/src/lib/prompts.ts:552]: Which ceremony would you like to finalize?

**Review by ctrlc03 — APPROVED** [review]
🚀 great work

File: packages/backend/src/functions/circuit.ts (hunk 2 above)
Comment (0xjei) [reply packages/backend/src/functions/circuit.ts:715]: Yep, will add

File: packages/phase2cli/src/commands/finalize.ts [hunk 5]
```diff
@@ -41,236 +51,334 @@ import {
...
+
+    // Prepare local paths.
+    const finalZkeyLocalFilePath = getFinalZkeyLocalFilePath(`${circuitPrefix}_${finalContributionIndex}.zkey`)
+    const verificationKeyLocalFilePath = getVerificationKeyLocalFilePath(`${circuitPrefix}_vkey.json`)
+    const verifierContractLocalFilePath = getVerifierContractLocalFilePath(`${circuitPrefix}_verifier.sol`)
+
+    // Prepare storage paths.
+    const verificationKeyStorageFilePath = getVerificationKeyStorageFilePath(
+        circuitPrefix,
+        `${circuitPrefix}_vkey.json`
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/finalize.ts:196]: common terms

File: packages/phase2cli/src/commands/finalize.ts [hunk 6]
```diff
@@ -41,236 +51,334 @@ import {
...
+
+    await sleep(2000) // workaound for descriptors.
+
+    // Extract data.
+    const { prefix: circuitPrefix } = circuit.data
+    const { prefix: ceremonyPrefix } = ceremony.data
+
+    // Prepare local paths.
+    const finalZkeyLocalFilePath = getFinalZkeyLocalFilePath(`${circuitPrefix}_${finalContributionIndex}.zkey`)
+    const verificationKeyLocalFilePath = getVerificationKeyLocalFilePath(`${circuitPrefix}_vkey.json`)
```
Comment (ctrlc03) [inline packages/phase2cli/src/commands/finalize.ts:190]: common terms

**Pull request merged**